Basic CRUD REST Endpoints for units in content libraries [FC-0083]

bradenmacdonald opened the pull request:
## Description

Implements https://github.com/openedx/frontend-app-authoring/issues/1703

This PR adds python and REST APIs for creating Units in content libraries.

## Supporting information

Depends on:
- https://github.com/openedx/opaque-keys/pull/369
- https://github.com/openedx/openedx-learning/pull/287
- https://github.com/openedx/openedx-events/pull/479

Includes commits from:
- https://github.com/openedx/edx-platform/pull/36381 (once it merges, this PR can be rebased to drop those changes)

## Testing instructions

1. Set up the dependency PRs in your devstack/python environment.
   - Tip: if you add `opaque-keys` to [this list](https://github.com/overhangio/tutor/blob/3f53104316761df4de4c526bfe72477e60050187/tutor/plugins/openedx.py#L109-L114) in `tutor` (we should upstream that fix), you can just run `tutor mounts add opaque-keys`, then from within your `cms` container make sure to run `pip install -e /mnt/opaque-keys` after checking out the correct branch, to refresh the entry points.
2. Run tests.
   - `tutor dev exec cms bash`
   - `DJANGO_SETTINGS_MODULE=cms.envs.tutor.test pytest openedx/core/djangoapps/content_libraries/ openedx/core/djangoapps/content/search/tests/ -x`

## Status

- [x] Create units
- [x] Set a "friendly" slug based on the title.
- [x] Read units
- [x] Generate a search index document for units
- [x] Build was entirely green as of 9383d5fde808dc49ac87e98bd8dba3ea308b2d36
- [x] Add units/containers to the search index automatically when a new unit is created
- [x] Include units/containers when rebuilding the entire search index via `reindex_studio` command.
- [x] Rebase after https://github.com/openedx/edx-platform/pull/36381 has merged
- [ ] Address TODOs added here, where possible
- [ ] Merge blocking PRs and update this accordingly

Out of scope for now: 
1. Update and Delete will be done in a follow-up PR: https://github.com/open-craft/edx-platform/pull/757
1. Anything involving children / components in the unit.

## Deadline

ASAP


Private ref: [FAL-4052](https://tasks.opencraft.com/browse/FAL-4052)

bradenmacdonald committed:
refactor: convert libraries API from attr.s to dataclass, fix types

bradenmacdonald committed:
feat: use new version of openedx-learning with containers support

openedx-webhooks commented:
Thanks for the pull request, @bradenmacdonald!


This repository is currently maintained by `@openedx/wg-maintenance-edx-platform`.


Once you've gone through the following steps feel free to tag them in a comment and let them know that your changes are ready for engineering review.

<details><summary>:radio_button: Get product approval</summary>

If you haven't already, [check this list](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Does-my-contribution-require-Product-Review%3F) to see if your contribution needs to go through the product review process.

- If it does, you'll need to submit a product proposal for your contribution, and have it reviewed by the [Product Working Group](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3449028609/Product+Working+Group).
    - This process (including the steps you'll need to take) is documented [here](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Product-Review-Process).
- If it doesn't, simply proceed with the next step.
</details>

<details><summary>:radio_button: Provide context</summary>

To help your reviewers and other members of the community understand the purpose and larger context of your changes, feel free to add as much of the following information to the PR description as you can:

- Dependencies
  > This PR must be merged before / after / at the same time as ...
- Blockers
  > This PR is waiting for OEP-1234 to be accepted.
- Timeline information
  > This PR must be merged by XX date **because** ...
- Partner information
  > This is for a course on edx.org.
- Supporting documentation
- Relevant [Open edX discussion forum](https://discuss.openedx.org/) threads
</details>





<details><summary>:radio_button: Get a green build</summary>

If one or more checks are failing, continue working on your changes until this is no longer the case and your build turns green.
</details><details>


---

<details><summary>Where can I find more information?</summary>

If you'd like to get more details on all aspects of the review process for open source pull requests (OSPRs), check out the following resources:

- [Overview of Review Process for Community Contributions](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/FAQ-about-pull-requests.html)
- [Pull Request Status Guide](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/pull-request-statuses.html)
- [Making changes to your pull request](https://docs.openedx.org/en/latest/documentors/how-tos/make_changes_to_your_pull_request.html)
</details>

<details><summary>When can I expect my changes to be merged?</summary>

Our goal is to get community contributions seen and reviewed as efficiently as possible.

However, the amount of time that it takes to review and merge a PR can vary significantly based on factors such as:

- The size and impact of the changes that it introduces
- The need for product review
- Maintenance status of the parent repository

:bulb: *As a result it may take up to several weeks or months to complete a review and merge your PR.*
</details>
<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

ChrisChV committed:
temp: Use opencraft branch of opaquekeys

ChrisChV committed:
refactor: Use LibraryElementKey instead of LibraryCollectionKey

bradenmacdonald committed:
refactor: split libraries API & REST API up into smaller modules

bradenmacdonald committed:
feat: new REST API for units in content libraries

bradenmacdonald committed:
fix: make corresponding updates to 'search' code

bradenmacdonald committed:
feat: python+REST API to get a unit

bradenmacdonald committed:
feat: auto-generate slug/key/ID from title of units

rpenido commented on openedx/core/djangoapps/content_libraries/rest_api/serializers.py:
```diff
@@ -230,6 +231,45 @@ class LibraryXBlockStaticFilesSerializer(serializers.Serializer):
+
+    # When creating a new container in a library, the slug becomes the ID part of
+    # the definition key and usage key:
+    slug = serializers.CharField(write_only=True)
```
Can we leave the `slug` optional so the frontend can only send the `display_name` when creating?
```suggestion
    slug = serializers.CharField(write_only=True, required=False)
```

rpenido replied:
Oops.. Already fixed! :smile:

bradenmacdonald replied:
😄

bradenmacdonald committed:
feat: generate search index documents for containers

pomegranited committed:
refactor: rename LibraryElementKey to LibraryItemKey

pomegranited committed:
fix: lint error

pomegranited committed:
Merge remote-tracking branch 'origin/master' into braden/units-api

pomegranited committed:
feat: adds new units to search index on create/update

and when running reindex_studio.

Updates requirements for openedx-events and openedx-learning to support
these changes.

rpenido committed:
fix: pylint

rpenido committed:
fix: temp requirement

rpenido committed:
fix: search index container events/tasks

rpenido commented on openedx/core/djangoapps/content/search/documents.py:
```diff
@@ -567,6 +567,7 @@ def searchable_doc_for_container(
         Fields.org: str(container_key.library_key.org),
         # In the future, this may be either course_container or library_container
         Fields.type: DocType.library_container,
+        Fields.block_type: container_key.container_type,
```
To check if it is "unit", "section", "subsection", etc..

pomegranited replied:
@rpenido  I added a comment to reflect this, but didn't change any code -- did you want me to?

rpenido replied:
I added this comment just to give context to my change :smile: 
Everything is fine.

rpenido committed:
Merge branch 'master' into braden/units-api

ChrisChV commented on openedx/core/djangoapps/content/search/documents.py:
```diff
@@ -546,3 +547,57 @@ def searchable_doc_for_collection(
+    doc = {
+        Fields.id: meili_id_from_opaque_key(container_key),
+        Fields.context_key: str(container_key.library_key),
+        Fields.org: str(container_key.library_key.org),
```
Nit, the `LibraryContainerLocator` has the [org property](https://github.com/openedx/opaque-keys/pull/369/files#diff-4c80f03bddd8f564f1b21a9b75defa1c134f277c4da5a4b1cb984f0c66f62280R1716)

```suggestion
        Fields.org: str(container_key.org),
```

pomegranited committed:
Merge remote-tracking branch 'origin/master' into braden/units-api

rpenido commented on openedx/core/djangoapps/content/search/api.py:
```diff
@@ -497,6 +523,22 @@ def index_collection_batch(batch, num_done, library_key) -> int:
+            units = authoring_api.get_units(library.learning_package_id)
+            num_units = units.count()
+            num_units_done = 0
+            status_cb(f"{num_units_done}/{num_units}. Now indexing units in library {lib_key}")
```
Should we loop directly through containers instead of units here to make it cleaner when we get to sections and subsections?

```suggestion
            containers = authoring_api.get_containers(library.learning_package_id)
            num_containers = units.count()
            num_containers_done = 0
            status_cb(f"{num_containers_done}/{containers_units}. Now indexing containers in library {lib_key}")
```

rpenido commented on openedx/core/djangoapps/content/search/api.py:
```diff
@@ -497,6 +523,22 @@ def index_collection_batch(batch, num_done, library_key) -> int:
+                    num_units_done,
+                    lib_key,
+                )
+                status_cb(f"{num_units_done}/{num_units} units indexed for library {lib_key}")
```
Ditto
```suggestion
                status_cb(f"{num_containers_done}/{num_containers} containers indexed for library {lib_key}")
```

rpenido commented on openedx/core/djangoapps/content/search/api.py:
```diff
@@ -475,6 +476,31 @@ def index_collection_batch(batch, num_done, library_key) -> int:
+                        library_key,
+                        container,
+                    )
+                    doc = searchable_doc_for_container(container_metadata.container_key)
```
Maybe we can avoid the `from_container` call here if we add a `get_library_container_usage_key` as we did here:

https://github.com/open-craft/edx-platform/blob/f2946bde3324a99fec8461e636b4ce91ae724423/openedx/core/djangoapps/content_libraries/api/libraries.py#L1712-L1720

I'm not sure we have the `container_type` on the `container` object (I think we should have?).

```suggestion
                    container_key = lib_api.get_library_container_usage_key(
                        library_key,
                        container.key,
                        container.type,  # ????????
                    )
```

Alternatively, we can pass the `container_metadata` along, similar as we have here:

https://github.com/open-craft/edx-platform/blob/8c6cbf02068955c823894c206ba9d2ad9f805630/openedx/core/djangoapps/content/search/documents.py#L491

pomegranited replied:
> I'm not sure we have the `container_type` on the container object (I think we should have?).

This is a good question.. @kdmccormick @ormsbee should the `Container` model have a `container_type` field (like `Component` has a `component_type`)? This would allow us to create the container locator/usage key from the `Container` object alone.

bradenmacdonald replied:
We don't have `container_type` on the container object, and technically it's not necessary because you can check which of the relationships exist - exactly one of `container.unit`, (or in the future) `container.subsection` or `container.section` will be set. If the container is loaded with `select_related` this is pretty efficient and I don't think we need the redundant `container_type` field, but we can always add it on later as an optimization.

rpenido reviewed (approved):
LGTM :+1:
Thank you for your work, @bradenmacdonald and @pomegranited!

- [x] I tested this using the instructions from the PR and creating a unit using https://github.com/openedx/frontend-app-authoring/pull/1741 
- [x] I read through the code
- [x] I checked for accessibility issues
- [x] Includes documentation

All my comments are non-blocking, in my opinion! :smile:

pomegranited committed:
feat: add get_library_container_usage_key to libraries API

and use it when search indexing containers

pomegranited committed:
fix: index all containers during reindex_studio

pomegranited committed:
chore: bump openedx-events requirement

pomegranited committed:
fix: address review comments

navinkarkera reviewed (approved):
@pomegranited @bradenmacdonald @rpenido Nice work!

- [x] I tested this: (Created new unit using https://github.com/openedx/frontend-app-authoring/pull/1741 and this PR)
- [x] I read through the code
- [x] I checked for accessibility issues
- [x] Includes documentation

pomegranited committed:
chore: bumps openedx-learning to 0.19.1

pomegranited committed:
Merge remote-tracking branch 'origin/master' into braden/units-api

kdmccormick reviewed (changes requested):
Over on the opaque-keys PR, I was assured that LibraryContainerLocators did not need to be UsageKeys 😄  But on this PR, I see that we're calling them usage_keys in various places.

We need to be consistent. Either these are not usage keys and we shouldn't call them as such, or they are usage keys and we should update opaque-keys to make LibraryContainerLocator explicitly subclass UsageKey.

ormsbee commented:
Would it be simpler to just use the container UUIDs in the REST API, instead of keys?

pomegranited committed:
fix: rename api method to library_container_locator

since container keys are locators, not usage keys

pomegranited commented:
@kdmccormick 

> Over on the opaque-keys PR, I was assured that LibraryContainerLocators did not need to be UsageKeys 😄 But on this PR, I see that we're calling them usage_keys in various places.
> 
> We need to be consistent.

Agreed -- please see [7081488](https://github.com/openedx/edx-platform/pull/36371/commits/70814889239ed340de95d406453f76a885e1c50b).

Note that I chose to leave the container search document [`usage_key` field](https://github.com/openedx/edx-platform/pull/36371/files#diff-8c7bc1adff4ab3b98ae2dfca83b4d50407bb2be71d45de93e504552424703fe2R572) named as such, so that its document format closely matches the [collection document](https://github.com/openedx/edx-platform/pull/36371/files#diff-8c7bc1adff4ab3b98ae2dfca83b4d50407bb2be71d45de93e504552424703fe2R529), which keeps the frontend code simpler. Let me know if that's not ok.

pomegranited commented:
@ormsbee 

> Would it be simpler to just use the container UUIDs in the REST API, instead of keys?

Currently, we're [slugifying the user's provided `title`](https://github.com/openedx/edx-platform/pull/36371/files#diff-5a248b18ae17c86d454776cb48e0372a8cbdc9610b3886f175a116f9b280e681R141-R143) and [passing it to `create_container` as `key=slug`](https://github.com/openedx/edx-platform/pull/36371/files#diff-5a248b18ae17c86d454776cb48e0372a8cbdc9610b3886f175a116f9b280e681R155). So that slug becomes the publishable entity key. I think that using this slug makes for a much nicer URL for the REST API than the UUID would, e.g. http://studio.local.openedx.io:8001/api/libraries/v2/containers/lct:OpenCraft:FAL-4054:unit:unit-1-4a4646/

If we're going to use the UUID in the LibraryContainerLocator instead of the slug, then I don't think there's any reason to slugify the title either, but will do it whichever way you think is best.

pomegranited committed:
chore: bumps opaque-keys dependency

pomegranited committed:
Merge remote-tracking branch 'origin/master' into braden/units-api

pomegranited committed:
Merge remote-tracking branch 'origin/master' into braden/units-api

pomegranited committed:
Merge remote-tracking branch 'origin/master' into braden/units-api

pomegranited commented:
Hi @kdmccormick and @ormsbee -- could you have another look at this PR to see if I've addressed your concerns here?

kdmccormick commented on openedx/core/djangoapps/content/search/tests/test_api.py:
```diff
@@ -208,6 +208,35 @@ def setUp(self):
+            "id": "lctorg1libunitunit-1-e4527f7c",
+            "block_id": "unit-1",
+            "block_type": "unit",
+            "usage_key": self.unit_usage_key,
```
```suggestion
            self.unit_key = "lct:org1:lib:unit:unit-1"
        self.unit_dict = {
            "id": "lctorg1libunitunit-1-e4527f7c",
            "block_id": "unit-1",
            "block_type": "unit",
            "usage_key": self.unit_key,
```

kdmccormick reviewed (approved):
Not a thorough review, but my concern is resolved, so 👍🏻 

If you want me to test and review thoroughly, let me know and I can do that!

pomegranited committed:
test: fix misnamed unit_usage_key

pomegranited commented:
Thanks @kdmccormick ! cf [e0f4380](https://github.com/openedx/edx-platform/pull/36371/commits/e0f4380f22c7df325607b0f93f2dcc193bdd0278)

> If you want me to test and review thoroughly, let me know and I can do that!

I'm happy with @rpenido and @navinkarkera 's reviews, but thank you for offering. I've got some follow-up work to do here in later PRs though, so if there's anything bugging you, please let me know :)

ormsbee commented:
@pomegranited: I'm good with trusting your existing reviewers on this. Thank you.

pomegranited committed:
feat: adds APIs to update or delete a container (#757)

* feat: adds python and REST APIs to update a container's display_name
* refactor: adds _get_container method to api to reduce code duplication
* feat: adds python and REST APIs to delete a container
* test: add container permission tests

pomegranited committed:
Merge remote-tracking branch 'origin/master' into braden/units-api
